Merge Subkey into sc-cli

seunlanlege opened the pull request:
see #4416

seunlanlege commented:
blocked on #4842

joepetrowski commented:
What is the motivation for this change? Would be nice to have a write up either here or in an issue.

NikVolf commented:
> What is the motivation for this change? Would be nice to have a write up either here or in an issue.

https://github.com/paritytech/substrate/issues/4416

cecton commented:
#4842 has been merged. Updating this PR is going to be fun :grin: call me if you need any help @seunlanlege

cecton commented on client/cli/src/arg_enums.rs:
```diff
@@ -82,6 +82,15 @@ arg_enum! {
 }
 
+arg_enum! {
+	#[allow(missing_docs)]
```
you don't need this "allow" as it is done on the module already

cecton replied:
you don't need this "allow" as it is done on the module already

cecton commented on client/cli/src/commands/runcmd.rs:
```diff
@@ -298,16 +275,16 @@ impl RunCmd {
+		} else if let Some(ref file) = self.shared_params.password_filename {
 			Some(fs::read_to_string(file).map_err(|e| format!("{}", e))?.into())
-		} else if let Some(ref password) = self.password {
+		} else if let Some(ref password) = self.shared_params.password {
```
If you moved some params to shared_params, you also need to move this code in the `update_config` of shared_params.

seunlanlege replied:
honestly not even sure what `update_config` does or why it exists.

cecton replied:
sc-cli provides an API to build a CLI. This API will help you build a Configuration (sc-service) used to parametrize your blockchain. The current pattern of sc-cli is:

```rust
let mut config = Configuration::new();
cmd.init();
cmd.update_config(config);
cmd.run(config);
```

Where you first create a default sc_service::Configuration for your blockchain. Then init things specialized to the CLI itself (logging). Then update the sc_service::Configuration with the parameters of the CLI. Then run the node (or subcommand) using the configuration.

So `update_config` is always a method implemented on a structopt to update the Configuration with the parameters of that structopt.

cecton commented on client/cli/src/commands/utils.rs:
```diff
@@ -0,0 +1,279 @@
+pub type HashFor<R> = <<R as RuntimeAdapter>::Runtime as frame_system::Trait>::Hash;
+
+/// Runtime adapter for signing utilities
+pub trait RuntimeAdapter: Sized {
```
Hah! Funny, I'm also implementing something similar to this

cecton replied:
:thinking: I think this is great actually. It means the API of the CLI is getting correct.

cecton replied:
I wonder if this is correct though... @seunlanlege does it make sense that it is in the CLI? It looks like parameters you want for a blockchain. Maybe it would make more sense to have this in sc_service?

seunlanlege replied:
> It looks like parameters you want for a blockchain.

Hmm, not sure what you mean, the core param here is the `Runtime` and crypto types.

> Maybe it would make more sense to have this in sc_service.

You're referring to the `RuntimeAdapter` trait, right? I'm not sure if it fits in sc_service since this is primarily used for cli, or maybe i don't fully understand where you're coming from.  I'm actually considering moving it to `sc-client-api`

seunlanlege replied:
> Hah! Funny, I'm also implementing something similar to this.

cool! what for though?

cecton replied:
In sc-cli you should have things more related to building a CLI with substrate. While in sc-service you have things more related to gluing up your blockchain in general (runtime, network etc...) (@tomaka maybe correct me if I'm wrong).

> cool! what for though?

I'm currently implementing a trait in sc-cli to replace the VersionInfo and the `load_spec` function of sc-cli. A CLI binary will have to impl this new trait on their structopt and it will "enrich"/"augment" the structopt to have all the stuff needed for a substrate CLI (mostly version information, program name, authors, etc... and load_spec).

cecton replied:
> we might want to consider RuntimeAdapter something to be implemented directly on Runtime. This is very much runtime dependent, and looks a lot like the TransacationSubmitter 

I'm closing this conversation in favor of https://github.com/paritytech/substrate/pull/4954#discussion_r394155651

seunlanlege commented on bin/node-template/node/src/cli.rs:
```diff
@@ -9,3 +11,21 @@ pub struct Cli {
+		(
+			frame_system::CheckVersion::new(),
+			frame_system::CheckGenesis::new(),
+			frame_system::CheckEra::from(Era::Immortal),
```
I think this should come from the cli, but not sure what that should look like from the cli side, @gnunicorn thoughts?

gnunicorn replied:
we might want to consider `RuntimeAdapter` something to be implemented directly on `Runtime`. This is very much runtime dependent, and looks a lot like the `TransacationSubmitter` – maybe we can combine those, @tomusdrw !?!

gnunicorn replied:
Let's make a first attempt in this PR and get that merged. Let's tackle https://github.com/paritytech/substrate/issues/5330 on a wider angle with the other parties involved after.

cecton replied:
I'm not in favor of merging this as it as it will block my work on the sc-cli part (if we do want to move RuntimeAdapter out of sc-cli)

gnunicorn replied:
sorry, I might have phrased that badly. I meant: let's take a crack at what I proposed, just not wait for 5330 – and be aware we will clean that approach up soon.

seunlanlege commented on bin/node-template/node/src/cli.rs:
```diff
@@ -9,3 +11,21 @@ pub struct Cli {
+			frame_system::CheckEra::from(Era::Immortal),
+			frame_system::CheckNonce::from(index),
+			frame_system::CheckWeight::new(),
+			pallet_transaction_payment::ChargeTransactionPayment::from(0),
```
so should this.

seunlanlege commented on client/cli/src/commands/transfer.rs:
```diff
@@ -0,0 +1,120 @@
+	{
+		let password = get_password(&self.shared_params)?;
+		let nonce = IndexFor::<RA>::from_str(&self.index).map_err(|e| format!("{}", e))?;
+		let to = AddressFor::<RA>::decode(&mut self.to.as_bytes()).map_err(|e| format!("{}", e))?;
```
I'm not confident that this is the right way to decode `AccountId`.

seunlanlege commented:
will merge with master once reviewed.

seunlanlege commented:
also, we'd have to add an exception for `sc-cli` to the dependency rules.

gnunicorn commented:
> also, we'd have to add an exception for `sc-cli` to the dependency rules.

No, we won't do that. Either we are able to do this via `traits` shared in `primitives` or we have to move this specific part into the `/utils/`-sections (rather than in `client/cli` directly) and make it something optional (maybe only enabled by a feature-flag). But frame and client may not overlap directly.

seunlanlege commented on bin/utils/subkey/src/main.rs:
```diff
@@ -99,11 +99,11 @@ trait Crypto: Sized {
-						format_public_key::<Self>(public_key.clone()),
-						format_account_id::<Self>(public_key),
-						Self::ss58_from_pair(&pair),
+					         uri,
```
Yeah, so I was going to ask if these files (generally the entire subkey folder) should be deleted in this PR?

cecton commented on client/cli/src/runner.rs:
```diff
@@ -216,9 +218,11 @@ impl<C: SubstrateCli> Runner<C> {
+			// Subcommand::Transfer(cmd) => cmd.run::<RA>(),
 			Subcommand::SignTransaction(cmd) => cmd.run::<RA>(),
 			Subcommand::Insert(cmd) => cmd.run::<RA>(),
+			_ => unimplemented!()
```
You can also use `todo!()`, faster to type and identical

NikVolf commented on client/cli/src/commands/utils.rs:
```diff
@@ -0,0 +1,253 @@
+	($scheme:expr, $method:ident($($params:expr),*)) => {
+		with_crypto_scheme!($scheme, $method<>($($params),*))
+	};
+    ($scheme:expr, $method:ident<$($gen:ident),*>($($params:expr),*)) => {
```
spaces

seunlanlege commented on client/cli/src/commands/tests.rs:
```diff
@@ -0,0 +1,101 @@
+use crate::Error;
+
+fn new_test_ext() -> sp_io::TestExternalities {
+	let t = frame_system::GenesisConfig::default().build_storage::<Runtime>().unwrap();
```
so not sure how to test these with the forbidden depencies rule. @gnunicorn

seunlanlege commented on client/cli/src/commands/transfer.rs:
```diff
@@ -0,0 +1,128 @@
+			AccountIdFor::<RA>::from_ss58check(&self.to).ok()
+				.expect("Invalid SS58-check address given for account ID.")
+		};
+		let amount = BalanceFor::<RA>::from_str(&self.amount).map_err(|e| format!("{}", e))?;
```
should we accept hex-encoded balances?

seunlanlege commented on client/cli/src/commands/transfer.rs:
```diff
@@ -0,0 +1,128 @@
+			<BalanceFor<RA> as FromStr>::Err: Display,
+	{
+		let password = self.keystore_params.read_password()?;
+		let nonce = IndexFor::<RA>::from_str(&self.index).map_err(|e| format!("{}", e))?;
```
also here, should we accept hex-encoded account indexes?

bkchr commented on client/cli/src/params/shared_params.rs:
```diff
@@ -43,6 +45,37 @@ pub struct SharedParams {
 	/// By default, all targets log `info`. The global log level can be set with -l<level>.
 	#[structopt(short = "l", long = "log", value_name = "LOG_PATTERN")]
 	pub log: Option<String>,
+
+	/// network address format
+	#[structopt(
+		long,
+		value_name = "NETWORK",
+		possible_values = &Ss58AddressFormat::all_names()[..],
+		parse(try_from_str = Ss58AddressFormat::try_from),
+		case_insensitive = true,
+		default_value = "polkadot"
+	)]
+	pub network: Ss58AddressFormat,
+
+	/// output format
+	#[structopt(
+		long,
+		value_name = "FORMAT",
+		possible_values = &OutputType::variants(),
+		case_insensitive = true,
+		default_value = "Text"
+	)]
+	pub output_type: OutputType,
+
+	/// cryptography scheme
+	#[structopt(
```
These should not be part of `SharedParams`. How should I use `output_type` when running a normal node?

seunlanlege replied:
true, good catch 👌🏾

bkchr commented on client/cli/src/commands/mod.rs:
```diff
@@ -53,14 +82,41 @@ pub enum Subcommand {
+	SignTransaction(SignTransactionCmd),
+
+	/// Author and sign a Node pallet_balances::Transfer transaction with a given (secret) key.
+	Transfer(TransferCmd),
```
How does this work for chains who don't use the `pallet-balances`?

seunlanlege replied:
well the requirement for using the CLI is your runtime must implement pallet-balances, although I think the logical next step is put stuff like this behind a feature flag

bkchr replied:
If we really need this (not sure), it should be made optional in its own crate. Like the `benchmark` sub command for example.

joepetrowski replied:
Does it work now? Last I knew, `subkey transfer` was really out of date.

bkchr replied:
I would expect that it works after this pr.

expenses commented on bin/node/browser-testing/Cargo.toml:
```diff
@@ -12,7 +12,7 @@ libp2p = { version = "0.18.0", default-features = false }
 serde = "1.0.106"
 serde_json = "1.0.48"
-wasm-bindgen = { version = "0.2.60", features = ["serde-serialize"] }
+wasm-bindgen = { version = "0.2.62", features = ["serde-serialize"] }
```
To fix the node browser test on CI, you should revert this change and re-run `cargo update`.

seunlanlege commented:
@gnunicorn @bkchr this is finally ready :sweat_smile:

gnunicorn commented on client/cli/src/commands/generate.rs:
```diff
@@ -0,0 +1,95 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
I regret to inform you we are now under a new license scheme, which puts all of client under GPL3-Classpath... 

The new header has to be:

```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

```

gnunicorn commented on client/cli/src/commands/generate_node_key.rs:
```diff
@@ -0,0 +1,61 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.
```

gnunicorn commented on client/cli/src/commands/insert.rs:
```diff
@@ -0,0 +1,131 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.
```

gnunicorn commented on client/cli/src/commands/inspect.rs:
```diff
@@ -0,0 +1,88 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.
```

gnunicorn commented on client/cli/src/commands/key.rs:
```diff
@@ -0,0 +1,61 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.


```

gnunicorn commented on client/cli/src/commands/key.rs:
```diff
@@ -0,0 +1,61 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+use crate::{error, substrate_cli_subcommands};
```
missing module level docs...

```suggestion

//! Implementation of the `key` cli command

use crate::{error, substrate_cli_subcommands};
```

gnunicorn commented on client/cli/src/commands/sign.rs:
```diff
@@ -0,0 +1,88 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

```

gnunicorn commented on client/cli/src/commands/sign_transaction.rs:
```diff
@@ -0,0 +1,107 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.
```

gnunicorn commented on client/cli/src/commands/tests.rs:
```diff
@@ -0,0 +1,70 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.
```

gnunicorn commented on client/cli/src/commands/utils.rs:
```diff
@@ -0,0 +1,252 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.
```

gnunicorn commented on client/cli/src/commands/vanity.rs:
```diff
@@ -0,0 +1,234 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.
```

gnunicorn commented on client/cli/src/commands/verify.rs:
```diff
@@ -0,0 +1,117 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2018-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or 
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.
```

gnunicorn commented on utils/cli-utils/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "cli-utils"
```
I am afraid the name [`cli-utils` is already occupied on crates.io](https://crates.io/crates/cli-utils). But even if not, these are binding frame and substrate so the name and location should reflect that:
```suggestion
name = "substrate-frame-cli-utils"
```

And the crate should be located at `utils/frame/cli-utils`.

gnunicorn commented on utils/cli-utils/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+version = "2.0.0-dev"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
+license = "GPL-3.0"
```
Utils are Apache-2.0 in the new scheme
```suggestion
license = "Apache-2.0"
```

gnunicorn commented on utils/cli-utils/src/lib.rs:
```diff
@@ -0,0 +1,43 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2019-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: Apache-2.0

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

gnunicorn commented on utils/frame/pallet-balances-cli/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+version = "2.0.0-dev"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
+license = "GPL-3.0"
```
```suggestion
license = "Apache-2.0"
```

gnunicorn commented on utils/frame/pallet-balances-cli/src/lib.rs:
```diff
@@ -0,0 +1,167 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
```suggestion
// This file is part of Substrate.

// Copyright (C) 2019-2020 Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: Apache-2.0

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

gnunicorn commented on client/cli/Cargo.toml:
```diff
@@ -24,13 +24,23 @@ lazy_static = "1.4.0"
 app_dirs = "1.2.1"
 tokio = { version = "0.2.9", features = [ "signal", "rt-core", "rt-threaded" ] }
 futures = "0.3.4"
+futures01 = { package = "futures", version = "0.1.29" }
```
I assume this was added because the original command had it, but do we _really_ need a 0.1-Future here? And if so, shouldn't we at least use the 0.3 compat-feature-future instead?

gnunicorn commented on client/cli/src/commands/insert.rs:
```diff
@@ -0,0 +1,131 @@
+use structopt::StructOpt;
+use sp_core::{crypto::KeyTypeId, Bytes};
+use std::convert::TryFrom;
+use futures01::Future;
```
see above.

gnunicorn commented on client/cli/src/commands/utils.rs:
```diff
@@ -0,0 +1,252 @@
+						Public key (hex): {}\n  \
+						Account ID:       {}\n  \
+						SS58 Address:     {}",
+				         uri,
+				         format_seed::<Pair>(seed),
+				         format_public_key::<Pair>(public_key.clone()),
+				         format_account_id::<Pair>(public_key),
+				         pair.public().into().into_account().to_ss58check(),
```
spaces...
```suggestion
						uri,
						format_seed::<Pair>(seed),
						format_public_key::<Pair>(public_key.clone()),
						format_account_id::<Pair>(public_key),
						pair.public().into().into_account().to_ss58check(),
```

gnunicorn commented on client/cli/src/commands/utils.rs:
```diff
@@ -0,0 +1,252 @@
+						Public key (hex): {}\n  \
+						Account ID:       {}\n  \
+						SS58 Address:     {}",
+				         uri,
+				         if let Some(seed) = seed { format_seed::<Pair>(seed) } else { "n/a".into() },
+				         format_public_key::<Pair>(public_key.clone()),
+				         format_account_id::<Pair>(public_key),
+				         pair.public().into().into_account().to_ss58check(),
```
spaces...

```suggestion
						uri,
						if let Some(seed) = seed { format_seed::<Pair>(seed) } else { "n/a".into() },
						format_public_key::<Pair>(public_key.clone()),
						format_account_id::<Pair>(public_key),
						pair.public().into().into_account().to_ss58check(),
```

gnunicorn commented on client/cli/src/commands/utils.rs:
```diff
@@ -0,0 +1,252 @@
+						Public key (hex):   {}\n  \
+						Account ID:         {}\n  \
+						SS58 Address:       {}",
+				         uri,
+				         String::from(v),
+				         format_public_key::<Pair>(public_key.clone()),
+				         format_account_id::<Pair>(public_key.clone()),
+				         public_key.to_ss58check_with_version(v),
```
spaces ...
```suggestion
					uri,
					String::from(v),
					format_public_key::<Pair>(public_key.clone()),
					format_account_id::<Pair>(public_key.clone()),
					public_key.to_ss58check_with_version(v),
```

gnunicorn commented on client/cli/src/runner.rs:
```diff
@@ -26,7 +26,14 @@ use log::info;
 use sp_runtime::traits::{Block as BlockT, Header as HeaderT};
 use sp_utils::metrics::{TOKIO_THREADS_ALIVE, TOKIO_THREADS_TOTAL};
-use std::{str::FromStr, fmt::Debug, marker::PhantomData, sync::Arc};
+use cli_utils::{AccountIdFor, AddressFor, IndexFor, BalanceFor, CallFor, RuntimeAdapter};
```
err...

this means, sc-cli is now pulling in substrate-frame-cli-utils (see below), which means it pulls in frame-specifics. That is not okay. You must still be able to use the cli even if you are not building any frame-runtime. That might mean some cli-subkey-related features are not available, but that's alright – or even better, if these would be expressed as traits in primitives and could be implemented outside of frame ... maybe in a `sp-cli` set...

gnunicorn commented on client/cli/src/runner.rs:
```diff
@@ -176,14 +183,20 @@ impl<C: SubstrateCli> Runner<C> {
 	/// A helper function that runs a future with tokio and stops if the process receives the signal
 	/// `SIGTERM` or `SIGINT`.
-	pub fn run_subcommand<B, BC, BB>(self, subcommand: &Subcommand, builder: B) -> Result<()>
+	pub fn run_subcommand<RA, B, BC, BB>(self, subcommand: &Subcommand, builder: B) -> Result<()>
```
If I am not running a FRAME-based Runtime, the other commands must still be available. I am afraid this must be refactored to allow for that...

gnunicorn commented on utils/cli-utils/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+documentation = "https://docs.rs/cli-utils"
+
+[dependencies]
+frame-system = { version = "2.0.0-alpha.4", default-features = false, path = "../../frame/system" }
+pallet-balances = { version = "2.0.0-alpha.4", default-features = false, path = "../../frame/balances" }
+sp-runtime = { version = "2.0.0-alpha.4", default-features = false, path = "../../primitives/runtime" }
```
old versions.
```suggestion
frame-system = { version = "2.0.0-alpha.7", default-features = false, path = "../../frame/system" }
pallet-balances = { version = "2.0.0-alpha.7", default-features = false, path = "../../frame/balances" }
sp-runtime = { version = "2.0.0-alpha.7", default-features = false, path = "../../primitives/runtime" }
```

seunlanlege commented:
```bash
error[E0046]: not all trait items implemented, missing: `Pair`
   --> /builds/parity/substrate/primitives/application-crypto/src/lib.rs:240:3
    |
240 |         impl $crate::CryptoType for Public {}
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Pair` in implementation
    | 
   ::: /builds/parity/substrate/primitives/application-crypto/src/ed25519.rs:29:5
    |
29  |     crate::app_crypto!(super, ED25519);
    |     ----------------------------------- in this macro invocation
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
    = help: implement the missing item: `type Pair = Type;`
error[E0046]: not all trait items implemented, missing: `Pair`
   --> /builds/parity/substrate/primitives/application-crypto/src/lib.rs:433:3
    |
433 |         impl $crate::CryptoType for Signature {}
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Pair` in implementation
    | 
   ::: /builds/parity/substrate/primitives/application-crypto/src/ed25519.rs:29:5
    |
29  |     crate::app_crypto!(super, ED25519);
    |     ----------------------------------- in this macro invocation
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
    = help: implement the missing item: `type Pair = Type;`
error[E0046]: not all trait items implemented, missing: `Pair`
   --> /builds/parity/substrate/primitives/application-crypto/src/lib.rs:240:3
    |
240 |         impl $crate::CryptoType for Public {}
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Pair` in implementation
    | 
   ::: /builds/parity/substrate/primitives/application-crypto/src/sr25519.rs:29:5
    |
29  |     crate::app_crypto!(super, SR25519);
    |     ----------------------------------- in this macro invocation
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
    = help: implement the missing item: `type Pair = Type;`
error[E0046]: not all trait items implemented, missing: `Pair`
   --> /builds/parity/substrate/primitives/application-crypto/src/lib.rs:433:3
    |
433 |         impl $crate::CryptoType for Signature {}
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Pair` in implementation
    | 
   ::: /builds/parity/substrate/primitives/application-crypto/src/sr25519.rs:29:5
    |
29  |     crate::app_crypto!(super, SR25519);
    |     ----------------------------------- in this macro invocation
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
    = help: implement the missing item: `type Pair = Type;`
error[E0046]: not all trait items implemented, missing: `Pair`
   --> /builds/parity/substrate/primitives/application-crypto/src/lib.rs:240:3
    |
240 |         impl $crate::CryptoType for Public {}
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Pair` in implementation
    | 
   ::: /builds/parity/substrate/primitives/application-crypto/src/ecdsa.rs:28:5
    |
28  |     crate::app_crypto!(super, ECDSA);
    |     --------------------------------- in this macro invocation
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
    = help: implement the missing item: `type Pair = Type;`
error[E0046]: not all trait items implemented, missing: `Pair`
   --> /builds/parity/substrate/primitives/application-crypto/src/lib.rs:433:3
    |
433 |         impl $crate::CryptoType for Signature {}
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Pair` in implementation
    | 
   ::: /builds/parity/substrate/primitives/application-crypto/src/ecdsa.rs:28:5
    |
28  |     crate::app_crypto!(super, ECDSA);
    |     --------------------------------- in this macro invocation
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
    = help: implement the missing item: `type Pair = Type;`
error: aborting due to 6 previous errors
For more information about this error, try `rustc --explain E0046`.
error: could not compile `sp-application-crypto`.
To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
ERROR: Job failed: exit code 1
```

I'm honestly not sure what's wrong.

rakanalh commented:
@seunlanlege 

The problem is as follows:

Since you enabled `full-crypto` feature on `sp_core` in 5a0616b. This means that the trait `CryptoType` needs to have `Pair` defined when implementing this trait.

https://github.com/paritytech/substrate/blob/9891eabe2a27d24734d6c58fc91e437e04e9a7bd/primitives/core/src/crypto.rs#L929-L934

If you look in `sp_application_crypto`, you'll notice that implementations of `CryptoType` do **not** as they assume that `full-crypto` feature is off in `sp_core`.

https://github.com/paritytech/substrate/blob/9891eabe2a27d24734d6c58fc91e437e04e9a7bd/primitives/application-crypto/src/lib.rs#L433

In this case, you should just update `sp_application_crypto` above to define the pair type in every place that implements `CryptoType`.

seunlanlege commented:
@gnunicorn this is ready for merge

gnunicorn commented on utils/frame/frame-utilities-cli/src/lib.rs:
```diff
@@ -0,0 +1,71 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! frame-system CLI utilities
```
this should probably also be apache-2...

gnunicorn commented on utils/frame/frame-utilities-cli/src/utils.rs:
```diff
@@ -0,0 +1,51 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
```
apache-2 again...

gnunicorn commented on utils/cli-utils/src/lib.rs:
```diff
@@ -1,18 +1,19 @@
+// limitations under the License.
 
 //! Utilities for cli.
 #![cfg_attr(not(feature = "std"), no_std)]
```
I think it would be super useful to add some rust docs here on what this is and how to use it, maybe even some example code people could copy to enable the features for their runtime. what do you think?

gnunicorn commented on utils/frame/frame-utilities-cli/Cargo.toml:
```diff
@@ -17,8 +17,21 @@ structopt = "0.3.8"
+sc-rpc = { version = "2.0.0-dev", path = "../../../client/rpc" }
+sc-service = { version = "0.8.0-dev", default-features = false, path = "../../../client/service" }
+sc-client-api = { version = "2.0.0-dev", default-features = false, path = "../../../client/api" }
+futures01 = { package = "futures", version = "0.1.29" }
```
so we settled on the futures01 thing? Can't used newer futures in compat mode?

seunlanlege replied:
unfortunately, this call requires a tokio v0.1 runtime.

```rust
http::connect(&url)
    .and_then(|client: AuthorClient<H, H>| {
        client.insert_key(key_type, suri, public).map(|_| ())
    })
    .map_err(|e| {
        println!("Error inserting key: {:?}", e);
    })
```

gnunicorn replied:
It sounds like an oversight on our side, but one we can fix later- that you'd need a f01 runtime to use the rpc client...

tomusdrw replied:
`futures` in the right version are re-exported from `jsonrpc_core`, so you can just do `use jsonprpc_core::futures::Future` instead.

gnunicorn commented on utils/cli-utils/Cargo.toml:
```diff
@@ -10,14 +10,17 @@ description = "CLI utils"
     "frame-system/std",
     "pallet-balances/std",
     "sp-runtime/std",
+	"sp-core/std",
```
don't ask me why, just keeping it consistent.
```suggestion
    "sp-core/std",
```

gnunicorn commented:
@bkchr still unhappy? If so, what is the way forward?

bkchr commented on utils/frame/frame-utilities-cli/src/lib.rs:
```diff
@@ -0,0 +1,72 @@
+#[cfg(feature = "balances")]
+mod transfer;
+#[cfg(feature = "balances")]
+pub use	transfer::TransferCmd;
```
```suggestion
pub use transfer::TransferCmd;
```

bkchr commented on utils/frame/frame-utilities-cli/src/lib.rs:
```diff
@@ -0,0 +1,72 @@
+	use structopt::StructOpt;
+	use std::io::Read;
+
+    #[test]
```
Spaces..

bkchr commented on utils/frame-utils/src/lib.rs:
````diff
@@ -0,0 +1,82 @@
+///
+/// ```rust,ignore
+/// use runtime::{Runtime, SignedExtra};
+/// use frame_utils::{IndexFor};
````
```suggestion
/// use frame_utils::IndexFor;
```

bkchr commented on utils/frame-utils/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+[features]
+default = ["std"]
+std = [
+    "frame-system/std",
```
Spaces.

bkchr commented on utils/frame/frame-utilities-cli/src/transfer.rs:
```diff
@@ -0,0 +1,171 @@
+        let nonce = IndexFor::<RA>::from_str(&self.index).map_err(|e| format!("{}", e))?;
+        let to = if let Ok(data_vec) = decode_hex(&self.to) {
+            AccountIdFor::<RA>::try_from(&data_vec)
+                .expect("Invalid hex length for account ID; should be 32 bytes")
```
Expect?

bkchr commented on utils/frame/frame-utilities-cli/src/transfer.rs:
```diff
@@ -0,0 +1,171 @@
+                .expect("Invalid hex length for account ID; should be 32 bytes")
+        } else {
+            AccountIdFor::<RA>::from_ss58check(&self.to).ok()
+                .expect("Invalid SS58-check address given for account ID.")
```
Same

bkchr commented on bin/node-template/runtime/src/lib.rs:
```diff
@@ -412,3 +413,19 @@ impl_runtime_apis! {
 	}
 }
+
+impl SignedExtensionProvider for Runtime {
```
Why do we need this in `no_std`?

seunlanlege replied:
that's because it's a subtrait of `frame_system::Trait`

seunlanlege replied:
or is it super-trait?

```rust
pub trait SignedExtensionProvider: frame_system::Trait {
    /// Concrete SignedExtension type.
    type Extra: SignedExtension;

    /// build extras for inclusion in extrinsics
    fn construct_extras(nonce: IndexFor<Self>) -> Self::Extra;
}
```

bkchr replied:
`frame_system::Trait` is the super trait of `SignedExtensionProvider`

seunlanlege replied:
sub trait then.

bkchr commented on utils/frame/frame-utilities-cli/src/transfer.rs:
```diff
@@ -0,0 +1,171 @@
+
+    /// The signing account's transaction index.
+    #[structopt(long)]
+    index: String,
```
Why is that a `String`?

Take a look at `BlockNumber` in `params/mod.rs` to get an idea how this should be implemented. (`BlockNumber` should be renamed and generalized for this)

seunlanlege commented:
![Screenshot from 2020-05-25 14-22-19](https://user-images.githubusercontent.com/16494314/82816510-2e285e00-9e93-11ea-8d6d-5804c741e104.png)
classic off by one error :sweat_smile:

seunlanlege commented:
@bkchr could you please check this again?

bkchr commented on utils/frame/frame-utilities-cli/src/lib.rs:
```diff
@@ -39,7 +39,7 @@ mod tests {
 
-    #[test]
+	#[test]
     fn generate() {
```
Still spaces.

seunlanlege commented:
@bkchr @gnunicorn this should be good now?

seunlanlege commented:
Yes I did, there are also tests for the commands.

bkchr commented:
> Yes I did, there are also tests for the commands.

You did not test it.

```
cargo run --release -- transfer --amount 1000 --from 0xa2bc899a8a3b16a284a8cefcbc2dc48a687cd674e89b434fbbdb06f400979744 --to 0xa2bc899a8a3b16a284a8cefcbc2dc48a687cd674e89b434fbbdb06f400979744 --index 1 --password test

   Compiling node-runtime v2.0.0-rc2 (/home/bastian/projects/parity/substrate/bin/node/runtime)
   Compiling node-executor v2.0.0-rc2 (/home/bastian/projects/parity/substrate/bin/node/executor)
   Compiling node-rpc v2.0.0-rc2 (/home/bastian/projects/parity/substrate/bin/node/rpc)
   Compiling node-cli v2.0.0-rc2 (/home/bastian/projects/parity/substrate/bin/node/cli)
    Finished release [optimized] target(s) in 1m 53s
     Running `target/release/substrate transfer --amount 1000 --from 0xa2bc899a8a3b16a284a8cefcbc2dc48a687cd674e89b434fbbdb06f400979744 --to 0xa2bc899a8a3b16a284a8cefcbc2dc48a687cd674e89b434fbbdb06f400979744 --index 1 --password test`
2020-05-29 13:54:44 🙇 You are using the peer ID QmaGKGpdm2iLiVCAnEcwrAhHxrcjMdGao4UubJxq7AF77n. This peer ID uses a legacy, deprecated representation that will no longer be supported in the future. Please refresh it by performing a RPC query to the appropriate node, by looking at its logs, or by using `subkey inspect-node-key` on its private key.
2020-05-29 13:54:44 🙇 You are using the peer ID QmaGKGpdm2iLiVCAnEcwrAhHxrcjMdGao4UubJxq7AF77n. This peer ID uses a legacy, deprecated representation that will no longer be supported in the future. Please refresh it by performing a RPC query to the appropriate node, by looking at its logs, or by using `subkey inspect-node-key` on its private key.
2020-05-29 13:54:44 🙇 You are using the peer ID QmbS1xrmWyP5h4xsPwAS7CJEJNStsuAUDUzhggqDJtLpou. This peer ID uses a legacy, deprecated representation that will no longer be supported in the future. Please refresh it by performing a RPC query to the appropriate node, by looking at its logs, or by using `subkey inspect-node-key` on its private key.
2020-05-29 13:54:44 🙇 You are using the peer ID QmbS1xrmWyP5h4xsPwAS7CJEJNStsuAUDUzhggqDJtLpou. This peer ID uses a legacy, deprecated representation that will no longer be supported in the future. Please refresh it by performing a RPC query to the appropriate node, by looking at its logs, or by using `subkey inspect-node-key` on its private key.
2020-05-29 13:54:44 🙇 You are using the peer ID QmdSHZLmwEL5Axz5JvWNE2mmxU7qyd7xHBFpyUfktgAdg7. This peer ID uses a legacy, deprecated representation that will no longer be supported in the future. Please refresh it by performing a RPC query to the appropriate node, by looking at its logs, or by using `subkey inspect-node-key` on its private key.
2020-05-29 13:54:44 🙇 You are using the peer ID QmdSHZLmwEL5Axz5JvWNE2mmxU7qyd7xHBFpyUfktgAdg7. This peer ID uses a legacy, deprecated representation that will no longer be supported in the future. Please refresh it by performing a RPC query to the appropriate node, by looking at its logs, or by using `subkey inspect-node-key` on its private key.
2020-05-29 13:54:44 🙇 You are using the peer ID Qmeo7it7YGbhGZqkDGpgnR3xMRFSp6AutA9oTDZPQyny8h. This peer ID uses a legacy, deprecated representation that will no longer be supported in the future. Please refresh it by performing a RPC query to the appropriate node, by looking at its logs, or by using `subkey inspect-node-key` on its private key.
2020-05-29 13:54:44 🙇 You are using the peer ID Qmeo7it7YGbhGZqkDGpgnR3xMRFSp6AutA9oTDZPQyny8h. This peer ID uses a legacy, deprecated representation that will no longer be supported in the future. Please refresh it by performing a RPC query to the appropriate node, by looking at its logs, or by using `subkey inspect-node-key` on its private key.

====================

Version: 2.0.0-rc2-91301561e0-x86_64-linux-gnu

   0: sp_panic_handler::set::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at /rustc/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libstd/panicking.rs:481
   2: rust_begin_unwind
             at /rustc/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libstd/panicking.rs:385
   3: core::panicking::panic_fmt
             at /rustc/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libcore/panicking.rs:89
   4: core::option::expect_failed
             at /rustc/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libcore/option.rs:1264
   5: sp_io::storage::get_version_1
   6: sp_io::storage::get
   7: frame_support::storage::unhashed::get
   8: <node_runtime::Runtime as frame_utils::SignedExtensionProvider>::construct_extras
   9: frame_utilities_cli::utils::create_extrinsic_for
  10: sc_cli::runner::Runner<C>::sync_run
  11: node_cli::command::run
  12: substrate::main
  13: std::rt::lang_start::{{closure}}
  14: std::rt::lang_start_internal::{{closure}}
             at /rustc/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libstd/rt.rs:52
      std::panicking::try::do_call
             at /rustc/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libstd/panicking.rs:297
      std::panicking::try
             at /rustc/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libstd/panicking.rs:274
      std::panic::catch_unwind
             at /rustc/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libstd/panic.rs:394
      std::rt::lang_start_internal
             at /rustc/2454a68cfbb63aa7b8e09fe05114d5f98b2f9740/src/libstd/rt.rs:51
  15: main
  16: __libc_start_main
  17: <unknown>
             at /build/glibc-2.30/csu/../sysdeps/x86_64/start.S:120


Thread 'main' panicked at '`get_version_1` called outside of an Externalities-provided environment.', primitives/io/src/lib.rs:79

This is a bug. Please report it at:

        https://github.com/paritytech/substrate/issues/new
```

seunlanlege commented:
ah didn't test `transfer` manually, but for the tests, I had to set up an externalities environment :thinking:, Is there some way to provide the cli with an externalities environment?

bkchr commented:
I already said that this was broken one week ago: https://github.com/paritytech/substrate/pull/4954#pullrequestreview-417062442

And said some words about it.

And your code also forces you to pass a password with `--password`, which makes no sense at all.

The old subkey code is also not deleted in this pr. Why isn't that done?

seunlanlege commented:
> I already said that this was broken one week ago: https://github.com/paritytech/substrate/pull/4954#pullrequestreview-417062442
> 
> And said some words about it.

Unfortunately didn't understand the comment. 

> And your code also forces you to pass a password with `--password`, which makes no sense at all.

The code is lifted from sub key but I guess we could remove this restriction. 

> The old subkey code is also not deleted in this pr. Why isn't that done?

From what @gnunicorn told me, subkey would still be shipped as a separate binary for now.

gnunicorn commented:
> From what @gnunicorn told me, subkey would still be shipped as a separate binary for now.

There must be a misunderstanding here: yes, `subkey` would still be shipped as an external binary, but it would reuse the existing cli-code rather than have a copy of it. We clearly shouldn't have to maintain the same code twice, and as the new one is generic, it shouldn't be too hard to implement the existing subkey functionality with it.

bkchr commented:
> Unfortunately didn't understand the comment at the time.

If you don't understand it, it is your task to ask and not just to ignore.

seunlanlege commented:
> If you don't understand it, it is your task to ask and not just to ignore.

Will definitely keep this in mind for next time ✊🏾. 

Meanwhile 🤔

> Is there some way to provide the cli with an externalities environment?

bkchr commented:
Having the externalities will not really help you, as I have written here:

> The next problem is also, that this transfer command will only work when native-version == runtime-version, because otherwise it can happen that the storage access will not work because stuff is now under different keys etc.

seunlanlege commented:
I don't really think that's a problem, we can do a check in the cli and let the user know that this command is unavailable until they update.

bkchr commented:
https://github.com/paritytech/substrate/blob/master/primitives/state-machine/src/read_only.rs#L40

seunlanlege commented:
@bkchr so I've removed the need for `SignedExtensionProvider::construct_extras` to access state by providing `AdditionalSigned` data manually as well as reimplemented subkey using the new `frame-utilities-cli`, `sc_cli` and `node_runtime`

gnunicorn commented on bin/node/runtime/src/lib.rs:
```diff
@@ -536,28 +537,34 @@ parameter_types! {
-			// The `System::block_number` is initialized with `n+1`,
-			// so the actual block number is `n`.
-			.saturating_sub(1);
+	fn construct_extras(index: IndexFor<Self>, era: Era, prior_block_hash: Option<Hash>) -> (
+		Self::Extra,
+		Option<<Self::Extra as SignedExtension>::AdditionalSigned>
+	) {
```
Reading through the list of parameters and the way this is used I am wondering if this is the best way to structure the needed parameters. Just consider that any new chain might add further information or – as we do here – there is another parameter `tip`, that isn't actually covered.

I wonder if having the input be an associated type as well and constraint the setting of specific values via a trait might be more flexible and future proof. [Something like this maybe](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c4bfb69402cc1fb49405e13666073a6c)?

seunlanlege replied:
hmm maybe i'm not seeing it, but why would that be preferable over say just passing a struct `ConstructExtrasParams` (which I remember thinking to do, but it skipped my mind) to the `construct_extras` method, which contains all the things it needs to actually `construct_extras` :D.

gnunicorn replied:
I think we mean the same thing. I just wonder how to instantiate that other params-thing and still have it extensible. My current though it is, we might want the return of `construct_extras` to be a result, so that if thing are missing, we can fail appropriately.

seunlanlege replied:
oh I was thinking more like:

```rust
#[derive(Default)]
pub struct ConstructExtrasParams<T: frame_system::Trait> {
    pub index: Option<IndexFor<T>>,
    pub era: Option<Era>,
    pub tip: Option<u64>,
    pub prior_block_hash: Option<Hash>,
    // so we can extend later on, 
    __nonexhaustive: ()
}

```

gnunicorn commented on utils/frame/frame-utilities-cli/Cargo.toml:
```diff
@@ -0,0 +1,39 @@
+version = "2.0.0-rc2"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
+license = "GPL-3.0"
```
license mismatch

gnunicorn commented on utils/frame/frame-utilities-cli/Cargo.toml:
```diff
@@ -0,0 +1,39 @@
+[package]
+name = "frame-utilities-cli"
```
this includes both substrate-client and frame features, as it is the integration layer. Thus the naming convention is to call this `substrate-frame-`...

```suggestion
name = "substrate-frame-utilities-cli"
```

gnunicorn commented on utils/frame/frame-utilities-cli/Cargo.toml:
```diff
@@ -0,0 +1,39 @@
+sp-runtime = { version = "2.0.0-rc2", path = "../../../primitives/runtime" }
+structopt = "0.3.8"
+codec = { version = "1.3.0", package = "parity-scale-codec" }
+pallet-balances = { version = "2.0.0-rc2", path = "../../../frame/balances" }
+pallet-indices = { version = "2.0.0-rc2", path = "../../../frame/indices" }
```
I expect these two to be `optional = true` and only activated if the named features were activated.

gnunicorn commented on utils/frame/frame-utilities-cli/Cargo.toml:
```diff
@@ -0,0 +1,39 @@
+frame-system = { version = "2.0.0-rc2", path = "../../../frame/system" }
+
+[features]
+balances = []
```
I'd expect all pallets-dependencies to be optionally activatable through feature flags.
```suggestion
default = []
balances = ["pallet-balances"]
indices = ["pallet-indices"]
```

gnunicorn commented on utils/frame/frame-utilities-cli/src/inspect.rs:
```diff
@@ -0,0 +1,89 @@
+// This file is part of Substrate.
```
why is this command here and not in `sc-cli`? This isn't FRAME specific.

seunlanlege replied:
it's part of the key subcommand. and one of them (insert_key) is frame-specific.

gnunicorn commented on utils/frame/frame-utilities-cli/src/insert.rs:
```diff
@@ -0,0 +1,135 @@
+	/// Run the command
+	pub fn run<P>(&self) -> Result<(), Error>
+		where
+			P: SignedExtensionProvider,
```
Why does this need `SignedExtensionProvider` and `HashFor<P>`? I'd expect this command to be generic without `FRAME`

seunlanlege replied:
`AuthorClient` needs a `Hash` type, here we're using `frame_system::Trait::Hash`

gnunicorn commented on utils/frame/frame-utilities-cli/src/inspect_node_key.rs:
```diff
@@ -0,0 +1,84 @@
+// This file is part of Substrate.
```
shouldn't this be in `sc-cli`? There is nothing frame-specific afaics.

gnunicorn commented on utils/frame/frame-utilities-cli/src/key.rs:
```diff
@@ -0,0 +1,77 @@
+			KeySubcommand::GenerateNodeKey(cmd) => cmd.run(),
+			KeySubcommand::Generate(cmd) => cmd.run(),
+			KeySubcommand::InspectKey(cmd) => cmd.run(),
+			KeySubcommand::Insert(cmd) => cmd.run::<P>(),
```
if `insert` doesn't actually need `P`, this wouldn't either.

gnunicorn commented on utils/frame/frame-utilities-cli/src/module_id.rs:
```diff
@@ -0,0 +1,117 @@
+// This file is part of Substrate.
```
Not sure this is located here correctly either.

gnunicorn commented on utils/frame/frame-utilities-cli/src/transfer.rs:
```diff
@@ -0,0 +1,152 @@
+// This file is part of Substrate.
```
This should probably only be included if the balances-features is set.

seunlanlege commented on bin/node/runtime/src/lib.rs:
```diff
@@ -606,9 +608,92 @@ parameter_types! {
 	pub const StakingUnsignedPriority: TransactionPriority = TransactionPriority::max_value() / 2;
 }
 
+#[derive(Default)]
+pub struct ExtrasParams<T: frame_system::Trait> {
+	era: Option<Era>,
+	index: Option<T::Index>,
+	tip: Option<u128>,
+	prior_block_hash: Option<T::Hash>,
+}
```
no reason tbh, can be changed

seunlanlege commented on bin/node/runtime/src/lib.rs:
```diff
@@ -608,15 +608,61 @@ parameter_types! {
+		),
+		&'static str
+	> {
+		let ExtrasParams {
+			tip,
+			era,
+			index,
+			prior_block_hash
+		} = params;
+		let (era, index) = (
+			era.ok_or("era is required")?,
+			index.ok_or("index is required")?,
+		);
+		let tip = tip.unwrap_or(0);
```
>  I don't think this should be conflated but rather the trait should have distinct hashes for era and genesis – and from my nderstanding similarly as to era and index neither of them is actually optional and should be checked for the same way.

hmm, i'm not sure i follow :sweat_smile:

gnunicorn replied:
two things:
1. genesis hash and prior_block_hash are not the same and can't be conflated into one entry
2. though it is part of the "optional" extras, `gensis` hash isn't actually optional. We probably want it to fail early, the same way it is done for era and index.

seunlanlege replied:
ah true,

seunlanlege commented on utils/frame-utils/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "frame-utils"
```
yeah i agree, frame_support is a better place for it.

seunlanlege replied:
ended up moving it to `frame_system::extras` because `frame_system` depends on `frame_support` whereas `frame_utils` depends on `frame_system` and so, therefore, can't be in `frame_support`

tomusdrw commented on bin/node/runtime/src/lib.rs:
```diff
@@ -624,9 +628,85 @@ parameter_types! {
+			prior_block_hash
+		} = params;
+		let (era, index) = (
+			era.ok_or("era is required")?,
```
IMHO that's not a good API design, would be better to ensure with type that these parameters are set or use some sensible defaults (i.e. immortal era, and from-state index)

gnunicorn replied:
@tomusdrw can you elaborate? Do you mean they shouldn't error if they are missing but have a reasonable default? Or do you think the API of having an associated input-params is bad? If so, what would be the proposed change?

tomusdrw replied:
Yeah, sorry for the random comment, with no clarity. I was skimming through the PR and noticed this.

The main reason I raised the comment is because I think having a generic error that indicates malformed input parameters in a function called `construct_extras` is a bit hard to reason about. If you just look at the types in the trait it seems okay to do:
```rust
Runtime::construct_extras(Runtime::extension_params())
``` 
and I'd assume that the "construction" of extras failed, not that my parameters are malformed, hence most likely I'd just propagate this error up instead of handling it, so clearly my own logical mistake is not caught by the type system, but rather would only pop up in the runtime.

I think there are two different paths that we can take:
1. Provide defaults inside `construct_extras` or
2. Separate `ExtrasParams` into a `Builder` and constructed struct.

In both cases `construct_extras` becomes infallible.

I think I'd prefer option (2) even though it will require more code, because (1) is:
- not explicit, 
- might become more difficult to maintain in the future,
- or the usage will end up being confusing (by adding more defaults)

With option (2) I imagine that `type Params` would be traitified with something like:
```rust
trait Params {
  trait Builder;
}
trait SignedExtensionProvider {
fn extension_params() -> Self::Params::Builder { .. }
fn construct_extras(params: Self::Params)...
} 
```
Also since the struct now becomes a builder the `fn set_*` methods might now return `&mut Self` which would allow for nice chaining.

Builder would be terminated with:
```rust
fn build(self) -> Result<Params, ExplicitNontStringBuilderError>;
// or maybe
fn build_with_defaults(self) -> Params;
 ```

That makes the error explicit to the builder and it's params - now it's a bit unclear if `construct_extras` fail because of malformed input or some other internal state. 
This forces the caller to handle the errors before the `construct_extras` can be called.

Hope it makes more sense :)

bkchr commented on utils/frame/frame-utilities-cli/src/utils.rs:
```diff
@@ -0,0 +1,74 @@
+		SignedPayload::from_raw(call, extra, additional_signed)
+	} else {
+		SignedPayload::new(call, extra)
+			.map_err(|_| "Transaction validity error")?
```
Why is that a validity error?

seunlanlege replied:
also copied code from subkey.

bkchr commented on utils/frame/frame-utilities-cli/src/utils.rs:
```diff
@@ -0,0 +1,74 @@
+/// create an extrinsic for the runtime.
+pub fn create_extrinsic_for<Pair, P, Call>(
+	call: Call,
+	nonce:  IndexFor<P>,
```
```suggestion
	nonce: IndexFor<P>,
```

bkchr commented on utils/frame/frame-utilities-cli/src/utils.rs:
```diff
@@ -0,0 +1,74 @@
+};
+use sp_runtime::generic::Era;
+
+/// create an extrinsic for the runtime.
```
Capital letter and what are the parameters?

bkchr commented on utils/frame/frame-utilities-cli/src/transfer.rs:
```diff
@@ -0,0 +1,155 @@
+    to: AddressFor<P>,
+    nonce: IndexFor<P>,
+    amount: BalanceFor<P>,
+    prior_block_hash: P::Hash
```
```suggestion
    prior_block_hash: P::Hash,
```

bkchr commented on utils/frame/frame-utilities-cli/src/sign_transaction.rs:
```diff
@@ -0,0 +1,125 @@
+	pass: Option<&str>,
+	call: CallFor<P>,
+	nonce: IndexFor<P>,
+	hash: P::Hash
```
```suggestion
	hash: P::Hash,
```

bkchr commented on utils/frame/frame-utilities-cli/src/transfer.rs:
```diff
@@ -0,0 +1,155 @@
+    }
+}
+
+fn print_ext<Pair, P>(
```
Function already exists in the other file

seunlanlege replied:
yeah they should be unified

bkchr commented on utils/frame/frame-utilities-cli/src/transfer.rs:
```diff
@@ -0,0 +1,155 @@
+    #[structopt(long)]
+    to: String,
+
+    /// genesis hash, for signed extensions.
```
Capital letter.

bkchr commented on utils/frame/frame-utilities-cli/src/transfer.rs:
```diff
@@ -0,0 +1,155 @@
+use pallet_balances::Call as BalancesCall;
+use crate::utils::create_extrinsic_for;
+
+type Bytes = Vec<u8>;
```
Why do we need this?

seunlanlege replied:
structopt's custom parser only works with type-aliases due to some proc macro issues.

bkchr commented on test-utils/runtime/Cargo.toml:
```diff
@@ -61,6 +62,7 @@ wasm-builder-runner = { version = "1.0.5", package = "substrate-wasm-builder-run
 default = [
 	"std",
 ]
+indices-lookup = []
```
Nope, no such feature here. Please remove

bkchr commented on primitives/core/src/crypto.rs:
```diff
@@ -376,6 +376,12 @@ macro_rules! ss58_address_format {
 
 		impl Ss58AddressFormat {
+			/// names of all address formats
+			pub fn all_names() -> [&'static str; 0 $(+ { let _ = $number; 1})*] {
```
```suggestion
			pub fn all_names() -> &[&'static str] {
```

bkchr commented on primitives/core/src/crypto.rs:
```diff
@@ -376,6 +376,12 @@ macro_rules! ss58_address_format {
 		impl Ss58AddressFormat {
+			/// names of all address formats
+			pub fn all_names() -> [&'static str; 0 $(+ { let _ = $number; 1})*] {
+				[
```
```suggestion
				&[
```

bkchr commented on frame/system/src/extras.rs:
````diff
@@ -0,0 +1,109 @@
+/// Strongly typed interface for the `SignedExtensions` that should be included in extrinsics
+/// for them to valid for your runtime.
+///
+/// ```rust,ignore
````
Make it compile or it is almost useless, because it gets outdated really fast

seunlanlege replied:
> Make it compile

unfortunately, this is impossible as that requires a full runtime so instead, I removed it.

bkchr commented on frame/system/src/extras.rs:
```diff
@@ -0,0 +1,109 @@
+	/// sets the nonce
+	fn set_era(&mut self, era: Era);
+	/// sets the prior block hash
+	fn set_prior_block_hash(&mut self, hash: T::Hash);
```
What is the prior hash?

seunlanlege replied:
this is the additional signed data for `CheckEra`, the hash of the block that starts the era. will rename it and add docs so that it's clearer

bkchr commented on frame/balances/src/lib.rs:
```diff
@@ -193,7 +192,7 @@ pub trait Subtrait<I: Instance = DefaultInstance>: frame_system::Trait {
 	/// The balance of an account.
 	type Balance: Parameter + Member + AtLeast32Bit + Codec + Default + Copy +
-		MaybeSerializeDeserialize + Debug;
+		MaybeSerializeDeserialize + MaybeFromStr + Debug;
```
Please revert

bkchr commented on frame/balances/src/lib.rs:
```diff
@@ -171,17 +171,16 @@ use sp_runtime::{
 	/// The balance of an account.
 	type Balance: Parameter + Member + AtLeast32Bit + Codec + Default + Copy +
-		MaybeSerializeDeserialize + Debug;
+		MaybeSerializeDeserialize + MaybeFromStr + Debug;
```
Same.

Add the bound where you need it

seunlanlege commented on client/cli/src/commands/generate.rs:
```diff
@@ -0,0 +1,105 @@
+
+	#[allow(missing_docs)]
+	#[structopt(flatten)]
+	pub shared_params: SharedParams,
```
in order to impl `CliConfiguration` for it

seunlanlege replied:
shared params is a requirement for it

bkchr replied:
You can not just mark this as solved because `CliConfiguration` requires it! I know that `CliConfiguration` requires it. But this is a really shitty CLI experience. If you don't need this, you need to adapt `CliConfiguration`.

bkchr commented on primitives/core/src/crypto.rs:
```diff
@@ -356,7 +356,8 @@ lazy_static::lazy_static! {
 		/// A known address (sub)format/network ID for SS58.
-		#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+		#[derive(Copy, Clone, PartialEq, Eq)]
+		#[cfg_attr(feature = "std", derive(Debug))]
```
You mean `RuntimeDebug`?

seunlanlege commented on client/cli/src/commands/module_id.rs:
```diff
@@ -0,0 +1,118 @@
+			.try_into()
+			.map_err(|_| "Cannot convert argument to moduleid: argument should be 8-character string")?;
+
+		let account_id: AccountId32 = ModuleId(id_fixed_array).into_account();
```
well, then we'd have to move this back out. so it can be generic over the runtime

seunlanlege commented on client/cli/src/commands/utils.rs:
```diff
@@ -0,0 +1,222 @@
+
+/// generate a pair from suri
+pub fn pair_from_suri<P: Pair>(suri: &str, password: Option<&str>) -> P {
+	P::from_string(suri, password).expect("Invalid phrase")
```
no reason tbh, code copied from subkey

seunlanlege commented on bin/node/runtime/src/lib.rs:
```diff
@@ -644,23 +724,27 @@ impl<LocalCall> frame_system::offchain::CreateSignedTransaction<LocalCall> for R
+		input.set_era(era);
+		let SignedExtensionData { extra, additional } = match Runtime::construct_extras(input) {
+			Ok(d) => d,
+			Err(_) => return None,
```
best we can do is log it

seunlanlege commented on client/cli/src/commands/module_id.rs:
```diff
@@ -0,0 +1,118 @@
+		let seed = "sdad1fb7";
+
+		let cmd = ModuleIdCmd::from_iter(&["moduleid", seed]);
+		assert!(cmd.run().is_ok());
```
in an ideal world we'd pass `impl std::io::Write` to our cli commands to run tests on the output. But that would require locking stdout and passing it around. and i'm not sure if we don't already lock it within the codebase or what do you think?

seunlanlege commented on bin/utils/subkey/src/main.rs:
```diff
@@ -16,816 +16,71 @@
+	author = "Parity Team <admin@parity.io>",
+	about = "Utility for generating and restoring with Substrate keys",
+)]
+pub enum Subkey {
```
hmm, not sure what you mean

bkchr replied:
When we want to create a polkadot specific `subkey`, we will need to copy all this code in this main file.

seunlanlege commented:
@tomusdrw I've implemented your suggestion as `ExtrasParamsBuilder` please check it out and let me know your thoughts.

@bkchr I've addressed most of your comments please re-review :pray:

tomusdrw commented on test-utils/runtime/src/lib.rs:
```diff
@@ -388,6 +389,12 @@ impl From<frame_system::Event<Runtime>> for Event {
 	}
 }
 
+impl From<pallet_balances::Event<Runtime>> for Event {
```
You can probably use `decl_event` macro instead.

seunlanlege replied:
actually this code is no longer needed

tomusdrw commented on test-utils/runtime/src/lib.rs:
```diff
@@ -398,8 +405,19 @@ parameter_types! {
+	type AccountStore = frame_system::Module<Runtime>;
+}
+
+
```
```suggestion
```

gavofyork commented:
@seunlanlege needs a master merge. @tomusdrw please merge if you're happy with it.

bkchr commented on utils/frame/frame-utilities-cli/src/transfer.rs:
```diff
@@ -0,0 +1,136 @@
+/// The `transfer` command
+#[derive(Debug, StructOpt)]
+#[structopt(
+    name = "transfer",
```
Spaces all over in this file.

bkchr commented on client/cli/src/commands/insert.rs:
```diff
@@ -0,0 +1,130 @@
+		H: DeserializeOwned + Serialize + Send + Sync + 'static,
+{
+	rt::run(
+		http::connect(&url)
```
If I call this as part of a node, why can I not insert the key directly via the file system?

I mean:
`node key insert`

This requires that node already runs. For `subkey` alone this made sense, but not when I call this from the node CLI.

bkchr commented on client/cli/src/params/shared_params.rs:
```diff
@@ -43,6 +46,49 @@ pub struct SharedParams {
 
+/// Optional flag for specifying crypto algorithm
+#[derive(Debug, StructOpt)]
+pub struct CryptoSchemeFlag {
```
Why are these 3 new structs here in `shared_params`? They don't belong here?

bkchr commented on frame/system/src/extras.rs:
```diff
@@ -0,0 +1,78 @@
+	type ExtrasParams;
+
+	/// Sets the tip
+	fn set_tip(self, tip: u128) -> Self;
```
You are aware that not every chain is required to have tips?

bkchr commented on frame/system/src/extras.rs:
```diff
@@ -0,0 +1,78 @@
+	/// Sets the block hash for the start of the era this transaction is valid for.
+	/// this is an additional signed data and is therefore **optional**
+	/// only provide it you know `construct_extras` will be called outside of an externalities environment.
+	fn set_starting_era_hash(self, hash: T::Hash) -> Self;
```
Also we don't require every chain to check for tx mortality?

gnunicorn commented:
> As you probably have seen, it is extreme complicated to construct a transaction for a runtime. There are still problems, like you expect a tip, era etc. This is wrong and not correct. The addition of creating transactions was an error back in the time for subkey IMHO. 

This is an actual hard issue because of the generic flexible system we want to have, and 

I've asked around and can't find anyone using `subkey` to make transactions. It was probably just created for presentation purposes but has vastly outlived its usefulness in the way that it is today (and substrate has evolved since). Let's remove the entire transaction generation things for now (from this PR and subkey/node in general) and if anyone steps forward and asks about  it, point them to [subxt](https://github.com/paritytech/substrate-subxt/)

seunlanlege commented:
heyy, @bkchr @gnunicorn, I've addressed all grumbles please give this one final (hopefully) once over.

expenses commented on bin/utils/subkey/src/lib.rs:
```diff
@@ -0,0 +1,81 @@
+	};
+
+	Ok(())
+}
```
```suggestion
}

```

expenses replied:
(Missing newline at EOF)

expenses commented on client/cli/src/commands/vanity.rs:
```diff
@@ -16,77 +16,92 @@
+pub struct VanityCmd {
+	/// Number of keys to generate
+	#[structopt(long, short)]
+	number: String,
```
Is this param used at the moment? If not, I'd remove it.

expenses commented on client/cli/src/commands/vanity.rs:
```diff
@@ -16,77 +16,92 @@
+
+	/// Desired pattern
+	#[structopt(long)]
+	pattern: Option<String>,
```
Can this be a `String` instead?

cecton replied:
If the `--pattern` is not required then it must be an `Option<String>`. If it is required, then it must be a `String`.

expenses commented on client/cli/src/commands/vanity.rs:
```diff
@@ -16,77 +16,92 @@
-	if !desired.chars().all(validate_base58) {
-		return Err("Pattern can only contains valid characters in base58 \
-			(all alphanumeric except for 0, l, I and O)");
-	}
```
In fact, what would be better than using a `String` for this would be to use a wrapper type that performs these checks in `FromStr`, so that it fails in the CLI interface and not here.

seunlanlege commented:
>  `SharedParams` is added by all the CLI commands.

Heyyyyy sorry about this, my strategy for removing unused-`SharedParams` is by removing the `CliConfiguration` implementation, it somehow skipped my mind to do the same for sign, verify and vanity.

expenses commented on bin/node/cli/Cargo.toml:
```diff
@@ -100,7 +100,6 @@ node-executor = { version = "2.0.0-rc5", path = "../executor" }
 sc-cli = { version = "0.8.0-rc5", optional = true, path = "../../../client/cli" }
 frame-benchmarking-cli = { version = "2.0.0-rc5", optional = true, path = "../../../utils/frame/benchmarking-cli" }
 node-inspect = { version = "0.8.0-rc5", optional = true, path = "../inspect" }
-
```
I'd keep this break.

bkchr commented on client/cli/src/commands/verify.rs:
```diff
@@ -0,0 +1,103 @@
+	} else {
+		Pair::Public::from_string(uri)
+			.ok()
+			.expect("Invalid URI; expecting either a secret URI or a public URI.")
```
Why is this an expect?

seunlanlege replied:
good catch

bkchr commented on utils/frame/frame-utilities-cli/src/module_id.rs:
```diff
@@ -0,0 +1,109 @@
+
+impl CliConfiguration for ModuleIdCmd {
+	fn shared_params(&self) -> &SharedParams {
+		&self.shared_params
```
Still shared params.

rakanalh commented on client/cli/src/commands/inspect_node_key.rs:
```diff
@@ -0,0 +1,75 @@
+
+use crate::{Error, NetworkSchemeFlag};
+use std::fs;
+use libp2p::identity::{PublicKey, ed25519};
```
I am curious as to why we're using ed25519 specific types from `libp2p`.

rakanalh commented on client/cli/src/commands/inspect_node_key.rs:
```diff
@@ -0,0 +1,75 @@
+	pub fn run(&self) -> Result<(), Error> {
+		let mut file_content = hex::decode(fs::read(&self.file)?)
+			.map_err(|_| "failed to decode secret as hex")?;
+		let secret = ed25519::SecretKey::from_bytes(&mut file_content)
```
Is this command only usable for ed25519? would it make sense to generalize this so that i can also get the public key from an `sr25519` file for example? I saw the `CryptoScheme` earlier and i think it can be used here to receive an additional argument for the crypto scheme.

seunlanlege replied:
I think for now, we only support ed25519, see 

https://github.com/paritytech/substrate/blob/00791403e55a9d4fc6d6a1ab8aa2fe8bb383546f/client/cli/src/arg_enums.rs#L80-L86

rakanalh commented on client/cli/src/commands/tests.rs:
```diff
@@ -16,36 +16,27 @@
+use super::{SignCmd, VanityCmd};
+use structopt::StructOpt;
+
+#[test]
```
Can you take each of these 2 tests into their respective command modules?

rakanalh commented on client/cli/src/commands/tests.rs:
```diff
@@ -16,36 +16,27 @@
+	assert!(sign.run().is_ok());
+}
+
+#[test]
```
Same as above

gnunicorn commented:
@seunlanlege [test is failing](https://gitlab.parity.io/parity/substrate/-/jobs/623687#L2174) and it seems [polkadot pr needs updating](https://gitlab.parity.io/parity/substrate/-/jobs/623695#L599).

gnunicorn commented:
bot merge
